### PR TITLE
Replace cdms2 with netCDF4 in PrePARE and Python tests

### LINKED ISSUE #490: Builds breaking from cdms2 issue related to libnetcdf
CMOR CI checks and nightly builds have recently been breaking due to an issue with cdms2.  The libnetcdf library has recently been updated but the nightly version of cdms2 relies on the older version.  This has caused the Python tests to break since they use cdms2.  They generate the following error.

```
Traceback (most recent call last):
  File "Test/test_python_CMIP6_CV_externalvariables.py", line 21, in <module>
    import cdms2
  File "/home/circleci/project/workspace/test_linux_cmor/miniconda/envs/py3.6/lib/python3.6/site-packages/cdms2/__init__.py", line 8, in <module>
    from . import dataset
  File "/home/circleci/project/workspace/test_linux_cmor/miniconda/envs/py3.6/lib/python3.6/site-packages/cdms2/dataset.py", line 8, in <module>
    from . import Cdunif
ImportError: libnetcdf.so.13: cannot open shared object file: No such file or directory
```

According to @dnadeau4, there is a dependency issue with the OSX version of cdms and hdf5 concerning glibfortran.  This has prevented newer version of cdms2 from being built that uses the newer libnetcdf 4.6.3.  The current workaround is to build CMOR with the older libnetcdf 4.6.2.  However, we want to go back to using the latest version of libnetdf as soon as possible.

Any suggestions are welcomed.

### PR DESCRIPTION
**Pull request by mauzey1**
Resolves #490 

This PR replaces the use of cdms2 in PrePARE and some of the Python tests with netCDF4.

Note: The MacOS CircleCI checks failed due to an error when creating the conda environment.

```
MD5MismatchError: Conda detected a mismatch between the expected content and downloaded content
for url 'https://repo.anaconda.com/pkgs/main/osx-64/libcxx-4.0.1-h579ed51_0.tar.bz2'.
  download saved to: /Users/distiller/project/workspace/test_macos_cmor/miniconda/pkgs/libcxx-4.0.1-h579ed51_0.tar.bz2
  expected md5 sum: 28471619a3a79da80d9b5747f1ec0221
  actual md5 sum: 291e61684d014641092020bdb1acb096
```

However, the changes that triggered the failed tests (https://github.com/PCMDI/cmor/commit/9f56aa6bf71c22e54315239c2e2f705d28cafa6a) didn't change the CircleCI configuration script.  The previous commit (https://github.com/PCMDI/cmor/commit/8ee711286de435522aaa5933ac848f1095389e91) that did change the CircleCI and Travis scripts was passing tests.

### CODE REVIEW

**Comment by doutriaux1** [comment]
@muryanto1 this is the same error as you. I think a package got corrupted on their server

**Comment by doutriaux1** [comment]
@mauzey1 can we have some test with cdms and some w/o like that we test both ways

File: LibCV/PrePARE/PrePARE.py [hunk 1]
```diff
@@ -457,15 +457,15 @@ def ControlVocab(self, ncfile, variable=None, print_all=True):
         for attr in ['branch_time_in_child', 'branch_time_in_parent']:
             if attr in list(self.dictGbl.keys()):
                 self.set_double_value(attr)
-                if not isinstance(self.dictGbl[attr], numpy.float64):
+                if not numpy.issubdtype(self.dictGbl[attr], numpy.float64):
```
Comment (doutriaux1) [inline LibCV/PrePARE/PrePARE.py]: nice!

File: LibCV/PrePARE/PrePARE.py [hunk 2]
```diff
@@ -535,7 +535,7 @@ def ControlVocab(self, ncfile, variable=None, print_all=True):
         # -------------------------------------------------------------------
         varid = cmip6_cv.setup_variable(variable_cmor_entry,
                                         self.dictVar['units'],
-                                        self.dictVar['_FillValue'][0],
```
Comment (doutriaux1) [inline LibCV/PrePARE/PrePARE.py]: is this due to the switch from cdms2 to netcdf?

Comment (mauzey1) [reply LibCV/PrePARE/PrePARE.py]: cdms2 was returning values as numpy ndarrays even if they were just one value.  netCDF4 returns numpy scalars for single values.

**Comment by mauzey1** [comment]
@doutriaux1 Should we include cdms2 along side netCDF4, or just use cdms2 in the testing environment for cdms tests?

**Comment by mauzey1** [comment]
@doutriaux1 The tests that I previously modified to use netCDF4 instead of cdms2 now have test cases for both.  cdms2 is installed in the testing environment but not for the build.

**Comment by mauzey1** [comment]
@doutriaux1 
The test [test_python_max_variables.py](https://github.com/PCMDI/cmor/blob/490_replace_cdms2_with_netcdf4/Test/test_python_max_variables.py) is failing on the MacOS checks.  It's giving this error:
```
HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140735957398464:
  #000: H5F.c line 444 in H5Fcreate(): unable to create file
    major: File accessibilty
    minor: Unable to open file
  #001: H5Fint.c line 1364 in H5F__create(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #002: H5Fint.c line 1559 in H5F_open(): unable to open file: name = 'CMIP6/CMIP6/ISMIP6/PCMDI/PCMDI-test-1-0/piControl-withism/r3i1p1f1/Amon/ta/gn/v20190531/ta_Amon_PCMDI-test-1-0_piControl-withism_r3i1p1f1_gnaG4OFK13537.nc', tent_flags = 13
    major: File accessibilty
    minor: Unable to open file
  #003: H5FD.c line 734 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = 'CMIP6/CMIP6/ISMIP6/PCMDI/PCMDI-test-1-0/piControl-withism/r3i1p1f1/Amon/ta/gn/v20190531/ta_Amon_PCMDI-test-1-0_piControl-withism_r3i1p1f1_gnaG4OFK13537.nc', errno = 24, error message = 'Too many open files', flags = 13, o_flags = 602
    major: File accessibilty
    minor: Unable to open file
```
I suspect it is due to it creating a lot of open file handles faster than it can close them.

The Python 3.7 MacOS check is getting an error at line 153 in Test/test_python_joerg_3.py.
https://github.com/PCMDI/cmor/blob/36351d7e29ee9be37a22f884d05e783222ed1d91/Test/test_python_joerg_3.py#L148-L154
It produces this error:
```
Traceback (most recent call last):
  File "Test/test_python_joerg_3.py", line 153, in <module>
    print(cmor.close(ivar1, True))
  File "/Users/distiller/project/workspace/test_macos_cmor/miniconda/envs/py3.7/lib/python3.7/site-packages/CMOR-3.4.0-py3.7-macosx-10.9-x86_64.egg/cmor/pywrapper.py", line 944, in close
    return _cmor.close(var_id, 1, 0)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xaf in position 1: invalid start byte
make[1]: *** [test_a_python] Error 1
make: *** [test_python] Error 2
Exited with code 2
```
It seems that Python 3.7 on OSX is having an issue with CMOR returning a file name when closing a variable that was already closed.

**Comment by mauzey1** [comment]
@doutriaux1 Looking at the function [cmor_close_variable](https://github.com/PCMDI/cmor/blob/36351d7e29ee9be37a22f884d05e783222ed1d91/Src/cmor.c#L5991-L6193), the parts that set the parameter `file_name` are skipped if the variable is closed.  This will cause `_cmor.close` in pywrapper.py to return an uninitialized string.

Should we set file_name to an empty string if the variable is closed?  Or, should we throw an error or warning if attempting to close an already closed variable?

**Comment by mauzey1** [comment]
@taylor13 @durack1 What should be the behavior of `cmor_close(var_id)` if `var_id` is for an already closed variable, or if CMOR is closed?  Warnings or errors?  Should it still return a file name for the variable if it was closed with `preserve` set to true?

**Comment by taylor13** [comment]
In this case I think a "Warning" should be raised (not an "error").  If "preserve" is set to true, I think is would be nice to return the file name (even if the file had already been closed), but if this is difficult to do, then don't bother.  [We should probably document the behavior in the CMOR documentation.

**Comment by mauzey1** [comment]
I'm not sure about how to handle the [test_python_max_variables.py](https://github.com/PCMDI/cmor/blob/master/Test/test_python_max_variables.py) error.  It seems to stem from an open file descriptor limit on CircleCI's MacOS.  It runs fine on my Macbook.  All it seems to be doing is creating, writing, and closing a variable 600 times.  I will comment it out of the Makefile for the time being.

**Comment by mauzey1** [comment]
@doutriaux1 Do you think this pull request can be approved?  Should I put the changes for the inclusion of the Python tests in another PR?

**Comment by doutriaux1** [comment]
@mauzey1 how does `make test` knows about which tests need cdms2 and which do not?

**Comment by mauzey1** [comment]
@doutriaux1 There are some Python tests, such as [test_python_history.py](https://github.com/PCMDI/cmor/blob/0d7618ae1c5dc47022e2321629964ff5c0aaff1a/Test/test_python_history.py#L23-L30), that check if cdms2 is installed.  If not installed then the test will exit.

There are similar checks for cdms2 in the CMIP6 CV tests.  The code below will skip the test case if cdms2 is not installed.
https://github.com/PCMDI/cmor/blob/0d7618ae1c5dc47022e2321629964ff5c0aaff1a/Test/test_python_CMIP6_CV_trackingNoprefix.py#L64-L66

Since cdms2 is a Python libary, should we only rerun the Python tests after installing cdms2?

**Comment by doutriaux1** [comment]
@mauzey1 looks good.

**Pull request merged**